src/analysis/normed_space/normed_group_hom.lean

```diff
@@ -0,0 +1,358 @@
+
+instance : inhabited (normed_group_hom V₁ V₂) := ⟨0⟩
+
+lemma coe_inj ⦃f g : normed_group_hom V₁ V₂⦄ (h : (f : V₁ → V₂) = g) : f = g :=
+by cases f; cases g; cases h; refl
+
+/-- The identity as a continuous normed group hom. -/
+@[simps]
+def id : normed_group_hom V V :=
+{ bound' := ⟨1, λ v, show ∥v∥ ≤ 1 * ∥v∥, by rw [one_mul]⟩,
+  .. add_monoid_hom.id V }
+
+/-- The composition of continuous normed group homs. -/
+@[simps]
+def comp (g : normed_group_hom V₂ V₃) (f : normed_group_hom V₁ V₂) :
+  normed_group_hom V₁ V₃ :=
+{ bound' :=
+  begin
+    obtain ⟨Cf, Cf_pos, hf⟩ := f.bound,
+    obtain ⟨Cg, Cg_pos, hg⟩ := g.bound,
+    use [Cg * Cf],
+    assume v,
+    calc ∥g (f v)∥
+        ≤ Cg * ∥f v∥    : hg _
+    ... ≤ Cg * Cf * ∥v∥ : _,
+    rw mul_assoc,
+    exact mul_le_mul le_rfl (hf v) (norm_nonneg _) Cg_pos.le
+  end
+  .. g.to_add_monoid_hom.comp f.to_add_monoid_hom }
+
+/-- Addition of normed group homs. -/
+instance : has_add (normed_group_hom V₁ V₂) :=
+⟨λ f g,
+{ bound' :=
+  begin
+    obtain ⟨Cf, Cf_pos, hCf⟩ := f.bound,
+    obtain ⟨Cg, Cg_pos, hCg⟩ := g.bound,
+    use [Cf + Cg],
+    assume v,
+    calc ∥f v + g v∥
+        ≤ ∥f v∥ + ∥g v∥ : norm_add_le _ _
+    ... ≤ Cf * ∥v∥ + Cg * ∥v∥ : add_le_add (hCf _) (hCg _)
+    ... = (Cf + Cg) * ∥v∥ : by rw add_mul
+  end,
+  .. (f.to_add_monoid_hom + g.to_add_monoid_hom) }⟩
+
+/-- Opposite of a normed group hom. -/
+instance : has_neg (normed_group_hom V₁ V₂) :=
+⟨λ f,
+{ bound' :=
+  begin
+    obtain ⟨C, C_pos, hC⟩ := f.bound,
+    use C,
+    assume v,
+    calc ∥-(f v)∥
+        = ∥f v∥ : norm_neg _
+    ... ≤ C * ∥v∥ : hC _
+  end, .. (-f.to_add_monoid_hom) }⟩
+
+/-- Subtraction of normed group homs. -/
+instance : has_sub (normed_group_hom V₁ V₂) :=
+⟨λ f g,
+{ bound' :=
+  begin
+    simp only [add_monoid_hom.sub_apply, add_monoid_hom.to_fun_eq_coe, sub_eq_add_neg],
+    exact (f + -g).bound'
+  end,
+  .. (f.to_add_monoid_hom - g.to_add_monoid_hom) }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : normed_group_hom V₁ V₂) = 0 := rfl
+
+@[simp] lemma coe_neg (f : normed_group_hom V₁ V₂) : ⇑(-f) = -f := rfl
+
+@[simp] lemma coe_add (f g : normed_group_hom V₁ V₂) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma coe_sub (f g : normed_group_hom V₁ V₂) : ⇑(f - g) = f - g := rfl
+
+/-- Homs between two given normed groups form a commutative additive group. -/
+instance : add_comm_group (normed_group_hom V₁ V₂) :=
+by refine_struct
+{ .. normed_group_hom.has_add, .. normed_group_hom.has_zero,
+  .. normed_group_hom.has_neg, ..normed_group_hom.has_sub };
+{ intros, ext, simp [add_assoc, add_comm, add_left_comm, sub_eq_add_neg] }
+.
+
+lemma sum_apply {ι : Type*} (s : finset ι) (f : ι → normed_group_hom V₁ V₂) (v : V₁) :
+  (∑ i in s, f i) v = ∑ i in s, (f i v) :=
+begin
+  classical,
+  apply finset.induction_on s,
+  { simp only [coe_zero, finset.sum_empty, pi.zero_apply] },
+  { intros i s his IH,
+    simp only [his, IH, pi.add_apply, finset.sum_insert, not_false_iff, coe_add] }
+end
```

Comment:
Can't you just reused the proof from `add_monoid_hom` here?

Reply:
What exactly do you have in mind?

Reply:
A proof that starts with `apply to_add_monoid_hom_injective`, and then can hopefully close the goal with `add_monoid_hom.finset_sum_apply`

Reply:
I don't think this works... because you need to know that a sum of `normed_group_hom`s is the same as the sum of add_monoid_homs or underlying functions. And that's basically what we are proving in this lemma.

Reply:
Ok, here's the proof I'm thinking of:
```lean
/-- Coercion of a normed_group_hom is an add_monoid_hom. Similar to `add_monoid_hom.coe_fn` -/
def coe_fn_add_hom : normed_group_hom V₁ V₂ →+ (V₁ → V₂) :=
{ to_fun := coe_fn, map_zero' := coe_zero, map_add' := coe_add}

@[simp] lemma coe_sum {ι : Type*} (s : finset ι) (f : ι → normed_group_hom V₁ V₂) :
  ⇑(∑ i in s, f i) = ∑ i in s, (f i) :=
(coe_fn_add_hom : _ →+ (V₁ → V₂)).map_sum f s

lemma sum_apply {ι : Type*} (s : finset ι) (f : ι → normed_group_hom V₁ V₂) (v : V₁) :
  (∑ i in s, f i) v = ∑ i in s, (f i v) :=
by simp
```
and then

Reply:
Ooh thanks! That's a nice idea.

Reply:
Fixed.